lib/tbot/tbot.go

```diff
@@ -118,109 +108,120 @@ func (b *Bot) Run(ctx context.Context) error {
 		return trace.Wrap(err)
 	}
 
-	// One-shot mode just invokes the output of credentials to the destinations.
-	// There's no retry logic here - this means we fail fast in the most common
-	// oneshot use-cases like CI-CD where backing off over several minutes on
-	// failure will just cost the customer money.
-	if b.cfg.Oneshot {
-		b.log.Info("One-shot mode enabled. Generating outputs.")
-		if err := b.renewOutputs(ctx); err != nil {
-			return trace.Wrap(err)
-		}
-
-		b.log.Info("Generated outputs. One-shot mode is enabled so exiting.")
-		return nil
-	}
-
-	// Warn about weird non-oneshot configuration.
-	if b.cfg.RenewalInterval > b.cfg.CertificateTTL {
-		b.log.Errorf(
-			"Certificate TTL (%s) is shorter than the renewal interval (%s). This is likely an invalid configuration. Increase the certificate TTL or decrease the renewal interval.",
-			b.cfg.CertificateTTL,
-			b.cfg.RenewalInterval,
-		)
-	}
+	// Create an error group to manage all the services lifetimes.
+	eg, egCtx := errgroup.WithContext(ctx)
+	services := []bot.Service{}
 
-	reloadBroadcast := channelBroadcaster{
+	// ReloadBroadcaster allows multiple entities to trigger a reload of
+	// all services. This allows os signals and other events such as CA
+	// rotations to trigger appropriate renewals.
+	reloadBroadcaster := &channelBroadcaster{
 		chanSet: map[chan struct{}]struct{}{},
 	}
-
-	// If in daemon mode, we spin up all of our separate concurrent components.
-	eg, egCtx := errgroup.WithContext(ctx)
-	eg.Go(func() error {
-		return trace.Wrap(b.caRotationLoop(egCtx, reloadBroadcast.broadcast))
-	})
-	eg.Go(func() error {
-		reloadCh, unsubscribe := reloadBroadcast.subscribe()
-		defer unsubscribe()
-		return trace.Wrap(b.renewBotIdentityLoop(egCtx, reloadCh))
-	})
-	eg.Go(func() error {
-		reloadCh, unsubscribe := reloadBroadcast.subscribe()
-		defer unsubscribe()
-		return trace.Wrap(b.renewOutputsLoop(egCtx, reloadCh))
-	})
+	// Trigger reloads from an configured reload channel.
 	if b.cfg.ReloadCh != nil {
-		eg.Go(func() error {
+		// We specifically do not use the error group here as we do not want
+		// this goroutine to block the bot from exiting.
+		go func() {
 			for {
 				select {
 				case <-egCtx.Done():
-					return nil
+					return
 				case <-b.cfg.ReloadCh:
-					reloadBroadcast.broadcast()
+					reloadBroadcaster.broadcast()
 				}
 			}
-		})
+		}()
 	}
 
+	b.botIdentitySvc = &identityService{
+		cfg:               b.cfg,
+		reloadBroadcaster: reloadBroadcaster,
+		log: b.log.WithField(
+			trace.Component, teleport.Component(componentTBot, "identity"),
+		),
+	}
+	// Initialize bot's own identity. This will load from disk, or fetch a new
+	// identity, and perform an initial renewal if necessary.
+	if err := b.botIdentitySvc.Initialize(ctx); err != nil {
```

Comment:
I agree - I considered making it something every service can do but I then started thinking about service dependencies, which felt like overkill. I figured I'd return to it when I had one which needed some kind of pre-start initialisation. 

I figure as well that this isn't too awful since I'm calling this against the concrete type, at this stage it's not /specifically/ a service.